In [1]:
import os
from time import time
from pathlib import Path
import pandas as pd 
import numpy as np
import torch
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from datasets import Dataset
import accelerate
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer

Obtenemos la carpeta actual:

In [2]:
current_dir = Path.cwd()
DATASETS_LOCATION = os.path.join(current_dir.parent.parent.parent, 'datasets')
MODELS_LOCATION = os.path.join(current_dir.parent.parent, 'models')

Cargamos el dataset en un dataframe:

In [3]:
df_disease_sp = pd.read_excel(os.path.join(DATASETS_LOCATION, 'disease_nlp_esp.xlsx'))
df_disease_sp.head(3)

Síntomas Enfermedad
0  Me siento cansado todo el tiempo y he notado q...     Anemia
1  Últimamente me falta el aire incluso después d...     Anemia
2  Durante las últimas semanas, me siento extrema...     Anemia

In [4]:
print(df_disease_sp['Enfermedad'].value_counts())


Enfermedad
Anemia       100
Talasemia    100
Trombosis    100
Diabetes     100
Name: count, dtype: int64


Observamos que ya las clases están bien distribuidas. Dividimos el dataset en conjunto de entrenamiento y prueba, y después dividieremos el conujunto de entrenamiento en conjunto de entrenamiento y conjunto de validación:

In [5]:
train_df, test_df = train_test_split(df_disease_sp, test_size=0.2, random_state=42, stratify=df_disease_sp['Enfermedad'])
train_df, val_df = train_test_split(train_df, test_size=0.2, random_state=42, stratify=train_df['Enfermedad'])

print("\nConjunto de entrenamiento:")
print(train_df['Enfermedad'].value_counts())

print("\nConjunto de validación:")
print(val_df['Enfermedad'].value_counts())

print("\nConjunto de prueba:")
print(test_df['Enfermedad'].value_counts())


Conjunto de entrenamiento:
Enfermedad
Trombosis    64
Anemia       64
Talasemia    64
Diabetes     64
Name: count, dtype: int64

Conjunto de validación:
Enfermedad
Trombosis    16
Talasemia    16
Anemia       16
Diabetes     16
Name: count, dtype: int64

Conjunto de prueba:
Enfermedad
Talasemia    20
Diabetes     20
Anemia       20
Trombosis    20
Name: count, dtype: int64


Codificamos las clases de 'Enfermedad', utilizando LabelEncoder:

In [6]:
label_encoder = LabelEncoder()

train_df['label'] = label_encoder.fit_transform(train_df['Enfermedad'])
val_df['label'] = label_encoder.transform(val_df['Enfermedad'])
test_df['label'] = label_encoder.transform(test_df['Enfermedad'])

train_df.head()

Síntomas Enfermedad  label
281  He notado que mi pierna está hinchada, y las v...  Trombosis      3
22   Estoy teniendo muchas infecciones, y mi piel s...     Anemia      0
146  Últimamente me siento muy débil, y mi piel est...  Talasemia      2
70   He notado que el blanco de mis ojos tiene un t...     Anemia      0
376  Tengo la vista borrosa, siempre tengo sed y pa...   Diabetes      1

## Entrenamiento

Definamos la función de métricas:

In [7]:
def compute_metrics(p):
    preds = np.argmax(p.predictions, axis=1)
    return {
        "accuracy": accuracy_score(p.label_ids, preds),
        "precision": precision_score(p.label_ids, preds, average='macro'),
        "recall": recall_score(p.label_ids, preds, average='macro'),
        "f1": f1_score(p.label_ids, preds, average='macro'),
    }

Definamos el modelo y el tokenizer. Utilizaremos el modelo preentrenado de hugging face 'BERTIN':

In [8]:
model_checkpoint = "bertin-project/bertin-roberta-base-spanish"  
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(model_checkpoint, num_labels=4)

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at bertin-project/bertin-roberta-base-spanish and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [9]:
# Preprocesamos los datos
def preprocess_function(examples):
    return tokenizer(examples['Síntomas'], 
                     padding=True, 
                     truncation=True,
                     return_tensors="pt")

Convertimos el dataframe en un dataset de hugging face, y lo tokenizamos:

In [10]:
train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)
validation_dataset = Dataset.from_pandas(val_df)

train_preprocessed_dataset = train_dataset.map(preprocess_function, batched=True)
test_preprocessed_dataset = test_dataset.map(preprocess_function, batched=True)
val_preprocessed_dataset = validation_dataset.map(preprocess_function, batched=True)

Map:   0%|          | 0/256 [00:00<?, ? examples/s]

Map:   0%|          | 0/80 [00:00<?, ? examples/s]

Map:   0%|          | 0/64 [00:00<?, ? examples/s]

In [11]:
training_args = TrainingArguments(
    output_dir="./results",
    evaluation_strategy="epoch",
    learning_rate=2e-5,              # Tasa de aprendizaje: se ha probado 5e-6, 1e-5, 5e-5 y 2e-5
    per_device_train_batch_size=8,   # Tamaño del batch para entrenamiento: se ha probado 8, 16, 32, 64
    per_device_eval_batch_size=8,    # Tamaño del batch para evaluación
    num_train_epochs=6,
    weight_decay=0.1,               # Decaimiento de peso: hemos probado 0.01, 0.1, 0.2, 0.3 y 0.4
    logging_dir="./logs",
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_preprocessed_dataset,
    eval_dataset=val_preprocessed_dataset,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics   # Función para calcular las métricas
)

c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\transformers\training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
C:\Users\maria\AppData\Local\Temp\ipykernel_13228\1012907193.py:12: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


## Training

Entrenamos nuestro modelo:

In [12]:
start = time()

trainer.train()

end = time()
print(f">>>>>>>>>>>>> elapsed time: {(end-start)/60:.0f}m")

  0%|          | 0/192 [00:00<?, ?it/s]

  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 1.0910271406173706, 'eval_accuracy': 0.75, 'eval_precision': 0.7873917748917749, 'eval_recall': 0.75, 'eval_f1': 0.7524846274846274, 'eval_runtime': 1.9322, 'eval_samples_per_second': 33.123, 'eval_steps_per_second': 4.14, 'epoch': 1.0}


  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 0.4518111050128937, 'eval_accuracy': 0.90625, 'eval_precision': 0.9131944444444444, 'eval_recall': 0.90625, 'eval_f1': 0.9079656862745098, 'eval_runtime': 1.8429, 'eval_samples_per_second': 34.728, 'eval_steps_per_second': 4.341, 'epoch': 2.0}


  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 0.22082951664924622, 'eval_accuracy': 0.96875, 'eval_precision': 0.9722222222222222, 'eval_recall': 0.96875, 'eval_f1': 0.9686274509803922, 'eval_runtime': 2.0242, 'eval_samples_per_second': 31.617, 'eval_steps_per_second': 3.952, 'epoch': 3.0}


  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 0.10690812021493912, 'eval_accuracy': 0.953125, 'eval_precision': 0.9605263157894737, 'eval_recall': 0.953125, 'eval_f1': 0.9527093596059113, 'eval_runtime': 2.0515, 'eval_samples_per_second': 31.196, 'eval_steps_per_second': 3.9, 'epoch': 4.0}


  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 0.07114793360233307, 'eval_accuracy': 0.96875, 'eval_precision': 0.9722222222222222, 'eval_recall': 0.96875, 'eval_f1': 0.9686274509803922, 'eval_runtime': 1.9353, 'eval_samples_per_second': 33.07, 'eval_steps_per_second': 4.134, 'epoch': 5.0}


  0%|          | 0/8 [00:00<?, ?it/s]

{'eval_loss': 0.11930215358734131, 'eval_accuracy': 0.953125, 'eval_precision': 0.9605263157894737, 'eval_recall': 0.953125, 'eval_f1': 0.9527093596059113, 'eval_runtime': 3.3806, 'eval_samples_per_second': 18.932, 'eval_steps_per_second': 2.366, 'epoch': 6.0}
{'train_runtime': 275.4887, 'train_samples_per_second': 5.576, 'train_steps_per_second': 0.697, 'train_loss': 0.3825710217158, 'epoch': 6.0}
>>>>>>>>>>>>> elapsed time: 5m


## Evaluación

Evaluamos nuestro modelo en el conjunto de prueba:

In [13]:
results = trainer.evaluate(test_preprocessed_dataset)

# Imprimimos las métricas de evaluación
print("Resultados de la evaluación en el conjunto de prueba:")
print(f"Accuracy: {results['eval_accuracy']:.4f}")
print(f"Precision: {results['eval_precision']:.4f}")
print(f"Recall: {results['eval_recall']:.4f}")
print(f"F1-Score: {results['eval_f1']:.4f}")

  0%|          | 0/10 [00:00<?, ?it/s]

Resultados de la evaluación en el conjunto de prueba:
Accuracy: 0.9750
Precision: 0.9756
Recall: 0.9750
F1-Score: 0.9750


Observamos que obtenemos una accuracy de 0.9531 en train, mientras que la accuracy es de 0.975 en test. Por tanto, nuestro modelo no presenta overfitting. Tal y como vemos, no se alcanza una accuracy de 1. Esto se debe a que las diferentes enfermedades que el modelo clasifica, tienen síntomas en común entre ellas, o entre algunas de ellas, haciendo difícil para el modelo poder distinguirlas basándose símplemente en los síntomas.

# Guardado del modelo

Ahora, guardamos el modelo y el tokenizer:

In [14]:
'''save_directory = os.path.join(MODELS_LOCATION, 'disease_classification_spanish_nlp')
model.save_pretrained(save_directory)
tokenizer.save_pretrained(save_directory)
print(f"Modelo y tokenizer guardados en el directorio {save_directory}")'''

Modelo y tokenizer guardados en el directorio c:\Users\maria\Desktop\universidad\master\TFM\tfm\src\models\disease_classification_spanish_nlp


Pasemos ahora a definir algunas frases de ejemplo para el test, que puedan confunir a nuestro modelo:

In [14]:
sample_sentences = [
    {"Síntomas": "Me siento cansado todo el tiempo, como si no tuviera energía para nada.", "Enfermedad": "Anemia"},
    {"Síntomas": "Mi hijo no crece como debería y siempre se queja de estar débil.", "Enfermedad": "Talasemia"},
    {"Síntomas": "Tengo dolores en los huesos y me siento agotado incluso después de descansar.", "Enfermedad": "Talasemia"},
    {"Síntomas": "Tengo mucha sed todo el tiempo y no dejo de ir al baño.", "Enfermedad": "Diabetes"},
    {"Síntomas": "Tengo un dolor punzante y la pierna está hinchada y roja.", "Enfermedad": "Trombosis"},
    {"Síntomas": "He bajado de peso sin razón y me siento muy cansado.", "Enfermedad": "Diabetes"}
]

# Convirtamos las frases de ejemplo a un dataframe
sample_df = pd.DataFrame(sample_sentences)

# Tokenicemos las frases de ejemplo
sample_dataset = Dataset.from_pandas(sample_df)
sample_preprocessed_dataset = sample_dataset.map(preprocess_function, batched=True)

# Efectuemos las predicciones
sample_predictions = trainer.predict(sample_preprocessed_dataset)
sample_preds = np.argmax(sample_predictions.predictions, axis=1)
sample_df["Enfermedad predicha"] = label_encoder.inverse_transform(sample_preds)

print("Resultados de las predicciones en las frases de ejemplo:")
sample_df

Map:   0%|          | 0/6 [00:00<?, ? examples/s]

  0%|          | 0/1 [00:00<?, ?it/s]

Resultados de las predicciones en las frases de ejemplo:


Síntomas Enfermedad  \
0  Me siento cansado todo el tiempo, como si no t...     Anemia   
1  Mi hijo no crece como debería y siempre se que...  Talasemia   
2  Tengo dolores en los huesos y me siento agotad...  Talasemia   
3  Tengo mucha sed todo el tiempo y no dejo de ir...   Diabetes   
4  Tengo un dolor punzante y la pierna está hinch...  Trombosis   
5  He bajado de peso sin razón y me siento muy ca...   Diabetes   

  Enfermedad predicha  
0            Diabetes  
1           Talasemia  
2            Diabetes  
3            Diabetes  
4           Trombosis  
5            Diabetes

# Test para comprobar que el modelo carga correctamente

In [16]:
model_directory = os.path.join(MODELS_LOCATION, 'disease_classification_spanish_nlp')
# Cargamos el tokenizer y el modelo
tokenizer = AutoTokenizer.from_pretrained(model_directory, local_files_only=True)
model_loaded = AutoModelForSequenceClassification.from_pretrained(model_directory, local_files_only=True)
def classify_disease_symptoms(input_text: str) -> str:
    prediction_mapping = {0: 'Anemia',
                          1: 'Diabetes', 
                          2: 'Talasemia', 
                          3: 'Trombosis'}
    # Tokenizamos la frase
    inputs = tokenizer(input_text, padding=True, truncation=True, return_tensors="pt")
    # Efectuamos las predicciones
    with torch.no_grad():
        outputs = model_loaded(**inputs)
        prediction = torch.argmax(outputs.logits, axis=1).item()
    return prediction_mapping[prediction]


# Añadimos una columna con las predicciones
def add_predictions(df):
    df['Predicted Enfermedad'] = df['Síntomas'].apply(classify_disease_symptoms)
    return df

# Actualizamos el dataframe con las predicciones
predicciones = add_predictions(test_df)
# Extraemos los valores reales y predichos
true_values = predicciones['Enfermedad']
predicted_values = predicciones['Predicted Enfermedad']

# Calculamos la accuracy
accuracy = accuracy_score(true_values, predicted_values)

# Mostramos el resultado
print(f"La accuracy del modelo es: {accuracy:.2f}")


La accuracy del modelo es: 0.97


Se observa que el modelo se ha cargado correctamente y que su rendimiento no se ha alterado.